File: Clients/PHP/phoenixDB.php
<?php

/**
 * First enable SOCKETS in php.ini file
 */

class PhoenixDB{
    /**
     * Set MAX_MESSAGE_LENGTH to accomodate upto 1MB of Data.
     */
    const MAX_MESSAGE_LENGTH = 1048576;
    private $socket;
    function __construct($host, $port, $username, $password, $database){
        
        /**
         * Prepare JSON String for authentication
         */
        $authData = json_encode(
            array(
                'username' => $username,
                'password' => $password,
                'database' => $database
            )
        );

        /**
         * Step 1: Create Socket
         * Step 2: Connect to Server
         * Step 3: Authenticate
         */
        $this->socket = socket_create(AF_INET, SOCK_STREAM, 0) or die("Failed to create Socket.");
        socket_connect($this->socket, $host, $port) or die("Failed to connect to $host:$port");
        socket_write($this->socket, $authData, strlen($authData));
    }

    function add($collection, $data){
        $query = json_encode(
            array(
                'operation' => 'add',
                'collection' => $collection,
                'data' => json_decode($data)
            )
        );
        socket_write($this->socket, $query, strlen($query));
        return socket_read($this->socket, self::MAX_MESSAGE_LENGTH);
    }

    function update($collection, $filter, $data){
        $query = json_encode(
            array(
                'operation' => 'update',
                'collection' => $collection,
                'filter' => json_decode($filter),
                'data' => json_decode($data)
            )
        );
        socket_write($this->socket, $query, strlen($query));
        return socket_read($this->socket, self::MAX_MESSAGE_LENGTH);
    }

    function find($collection, $filter){
        $query = json_encode(
            array(
                'operation' => 'find',
                'collection' => $collection,
                'filter' => json_decode($filter)
            )
        );
        socket_write($this->socket, $query, strlen($query));
        return socket_read($this->socket, self::MAX_MESSAGE_LENGTH);
    }

    function delete($collection, $filter){
        $query = json_encode(
            array(
                'operation' => 'delete',
                'collection' => $collection,
                'filter' => json_decode($filter)
            )
        );
        socket_write($this->socket, $query, strlen($query));
        return socket_read($this->socket, self::MAX_MESSAGE_LENGTH);
    }

    function __destruct() {
        socket_close($this->socket);
    }


}

?>

File: Clients/Java/PhoenixDB.java
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;

public class PhoenixDB {
    Socket socket = null;
    InputStream inputStream = null;
    OutputStream outputStream = null;

    int MAX_STREAM_SIZE = 256000;

    PhoenixDB(String host, int port, String username, String password, String database){
        try {

            /**
             * Connect to Database
             * */
            socket = new Socket(host, port);

            /**
             * Initialize Input Stream.
             * */
            inputStream = socket.getInputStream();

            /**
             * Initialize Output Stream
             * */
            outputStream = socket.getOutputStream();

            /**
             * Authenticate Username and Password and Connect and get Access to Database
             * */
            String auth = "{";
            auth += "\"username\":\""+username+"\",";
            auth += "\"password\":\""+password+"\",";
            auth += "\"database\":\""+database+"\"";
            auth += "}";

            /**
             * Send Auth data from Login
             */
            outputStream.write(auth.getBytes("UTF8"));
        } catch (IOException e) {
            System.out.println(e.getMessage());
        }
    }

    public String add(String data, String collection) {
        byte[] response = new byte[MAX_STREAM_SIZE];
        try{
            JSONParser jsonParser = new JSONParser();
            JSONObject dataObject = (JSONObject)jsonParser.parse(data);
            JSONObject command = new JSONObject();
            command.put("operation", "add");
            command.put("collection", collection);
            command.put("data", dataObject);
            outputStream.write(command.toJSONString().getBytes(StandardCharsets.UTF_8));
            inputStream.read(response);
        }catch (IOException e){
            System.out.println(e.getMessage());
            System.out.println("IOException");
        } catch (ParseException e){
            System.out.println("Parse Exception");
            System.out.println(e.getMessage());
        }
        return new String(trim(response), StandardCharsets.UTF_8);
    }

    public String find(String collection, String filter){
        byte[] response = new byte[MAX_STREAM_SIZE];
        try{
            JSONParser jsonParser = new JSONParser();
            JSONObject filterObject = (JSONObject)jsonParser.parse(filter);
            JSONObject command = new JSONObject();
            command.put("operation", "find");
            command.put("collection", collection);
            command.put("filter", filterObject);
            outputStream.write(command.toJSONString().getBytes(StandardCharsets.UTF_8));
            inputStream.read(response);
        }catch (IOException e){
            System.out.println(e.getMessage());
            System.out.println("IOException");
        } catch (ParseException e){
            System.out.println("Parse Exception");
            System.out.println(e.getMessage());
        }
        return new String(trim(response), StandardCharsets.UTF_8);
    }

    public String delete(String collection, String filter){
        byte[] response = new byte[MAX_STREAM_SIZE];
        try{
            JSONParser jsonParser = new JSONParser();
            JSONObject filterObject = (JSONObject)jsonParser.parse(filter);
            JSONObject command = new JSONObject();
            command.put("operation", "delete");
            command.put("collection", collection);
            command.put("filter", filterObject);
            outputStream.write(command.toJSONString().getBytes(StandardCharsets.UTF_8));
            inputStream.read(response);
        }catch (IOException e){
            System.out.println(e.getMessage());
            System.out.println("IOException");
        } catch (ParseException e){
            System.out.println("Parse Exception");
            System.out.println(e.getMessage());
        }
        return new String(trim(response), StandardCharsets.UTF_8);
    }

    public String update(String collection, String filter, String data){
        byte[] response = new byte[MAX_STREAM_SIZE];
        try{
            JSONParser jsonParser = new JSONParser();
            JSONObject filterObject = (JSONObject)jsonParser.parse(filter);
            JSONObject dataObject = (JSONObject)jsonParser.parse(data);

            JSONObject command = new JSONObject();
            command.put("operation", "update");
            command.put("collection", collection);
            command.put("filter", filterObject);
            command.put("data", dataObject);

            outputStream.write(command.toJSONString().getBytes(StandardCharsets.UTF_8));
            inputStream.read(response);
        }catch (IOException e){
            System.out.println(e.getMessage());
            System.out.println("IOException");
        } catch (ParseException e){
            System.out.println("Parse Exception");
            System.out.println(e.getMessage());
        }
        return new String(trim(response), StandardCharsets.UTF_8);
    }


    public void close(){
        try{
            socket.close();
            inputStream.close();
            outputStream.close();
        }catch (Exception e){
            System.out.println(e.getMessage());
        }
    }


    private static byte[] trim(byte[] bytes) {
        int i = bytes.length - 1;
        while (i >= 0 && bytes[i] == 0){
            i--;
        }
        return Arrays.copyOf(bytes, i + 1);
    }
}
